docs/design_notes/lift.md

```diff
@@ -0,0 +1,338 @@
+For example: `vmap` has broadcasted and vectorized arguments, while `scan` has scan, carry, and broadcast arguments.
+A variable would have to define its behavior for all these transformations otherwise a `Module` would not be compatible with
+these tranformations. Or alternatively, we would have to make default decisions for how transformations are handled.
+However, this could lead to silent bugs becuase the behavior might not actually be valid given the users intent. 
+
+The lift package also provides a general purpose `transform` which allows an arbitrary function to transforms a variable collection.
+This can be used to for example tie the weights in a tied autoencoder by transposing the weights.
+It is unclear whether a similair general purpose transform could be defined if lifting decisions were made at variable definition.
+
+
+### Linen
+
+The lifting module does not know about linen Modules. Instead it operates directly on instances of `flax.core.Scope`.
+A `Scope` contains the variables and PRNG sequences of a `Module`. Each module has a `Scope` instance linked to it as long
+as its bound to a parent or it was created using a `init` or `apply`.
+
+When a `Module` is transformed we use the `flax.core.lift` APIs to lift the scope and use `Module.clone()` to create a new `Module` instance with the lifted `Scope` bound to it.
+
+`flax.linen.transforms` exposes wrappers for the transformations in `flax.core.lift`. The core lifting APIs operate on functions while
+the linen wrappers can transform either a `Module` class or a `Module` method.
+
+Lifting is thus implemented independently from the linen API. This seperation of concern simplifies the implementation while also allowing other Module APIs that chooce a different syntatic sugar for defining Modules to reuse the very generic and flexible `Scope `+ lift APIs for handling state and RNGs. 
+
+
+### Implementation
+
+The `pack(fn, in_vars, out_vars, rngs)` API goes through the following stages:
+
+
+1. *Scope deduplication*
+
+    This stage is only relevant if multiple Scopes are lifted together.
+    In this case we must first find the set of root scopes.
+    A scope is a root if none of its ancestors is in the set of scopes that is too be lifted.
+
+    By only lifting roots we avoid lifting the same variables twice.
+
+    For non-root scopes we store a reference to its ancestor scope and a path such that we can later reconstruct it (stage 4).
+
+2. *Filter stage*
+
+    Variables & PRNG sequences are split up into groups. This way `fn` can lift each group into the seperation seperatly.
+    A group is defined by a filter which can be specified as:
+    - a list of collections/prng names
+    - `True` (match everything)
+    - `False` (match nothing)
+    - `DenyList(filter)` (match everything but the specified filter).
+
+    A collection or PRNG sequence can only be put into a single group. If a collection matches multiple filters it will be put into
```

Comment:
Nit:

```suggestion
    A collection or a PRNG sequence can only be put into a single group. If a collection matches multiple filters, it will be put into
```